### PR TITLE
Use new Maven artifacts from Gradle

### PR DESCRIPTION
**Pull request by blasten**
## Description

* Migrate Gradle, so it consumes the Maven artifacts uploaded to the new cloud bucket. 
* This PR also reduces a bit of technical debt in the build process--the build doesn't disassemble `flutter.jar` to construct fat APKs.

## Related Issues

* https://github.com/flutter/flutter/issues/11439
* https://github.com/flutter/flutter/issues/36499
* https://github.com/flutter/flutter/issues/34889
* https://github.com/flutter/flutter/issues/34444

* Fixes Proguard in release mode: https://github.com/flutter/flutter/issues/39187, https://github.com/flutter/flutter/issues/37441, https://github.com/flutter/flutter/issues/36160, https://github.com/flutter/flutter/issues/31309

## Tests

I added the following tests:

Unit tests in `cache_test.dart`.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by jonahwilliams — APPROVED** [review]
LGTM with caveat about naming android artifact

File: packages/flutter_tools/lib/src/commands/precache.dart [hunk 1]
```diff
@@ -15,8 +15,12 @@ class PrecacheCommand extends FlutterCommand {
         help: 'Precache artifacts for all host platforms.');
     argParser.addFlag('force', abbr: 'f', negatable: false,
         help: 'Force downloading of artifacts.');
-    argParser.addFlag('android', negatable: true, defaultsTo: true,
-        help: 'Precache artifacts for Android development.');
+    argParser.addFlag('android_gen_snapshot', negatable: true, defaultsTo: true,
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/precache.dart]: Are there situations where any of these would be needed without the others? Would it make sense to hide fine-grained options (`hide: !verbose`), and keep a non-hidden 'android' option?

Comment (blasten) [reply packages/flutter_tools/lib/src/commands/precache.dart]: We map each flag to individual `DevelopmentArtifact`. So, we would need another layer of indirection to map `android` to a group of `DevelopmentArtifact`.

Comment (zanderso) [reply packages/flutter_tools/lib/src/commands/precache.dart]: If you get the `--android` flag, you can split it apart into `android_gen_snapshot`, `android_maven`, and `android_internal_build`. `--android` can be shown in help by default, `--android_gen_snapshot`, `--android_maven`, and `--android_internal_build` can be hidden unless `--verbose` is also given by passing `hide: !verbose` to `argParser.addFlag()`.

This is only needed if using the new `--android_`* flags individually is uncommon or only used by people working on the tool.

**Review by zanderso — APPROVED** [review]

File: packages/flutter_tools/lib/src/cache.dart [hunk 2]
```diff
@@ -30,7 +32,11 @@ class DevelopmentArtifact {
   final bool unstable;
 
   /// Artifacts required for Android development.
-  static const DevelopmentArtifact android = DevelopmentArtifact._('android');
+  static const DevelopmentArtifact android_gen_snapshot = DevelopmentArtifact._('android_gen_snapshot');
+  static const DevelopmentArtifact android_maven = DevelopmentArtifact._('android_maven');
+  // Artifacts used for internal builds.
```
[the remark this answers is not in the document]
Comment (blasten) [reply packages/flutter_tools/lib/src/cache.dart]: Done 💯

File: packages/flutter_tools/lib/src/commands/precache.dart (hunk 1 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/commands/precache.dart]: 💯Done

**Pull request merged**